File: web/core/casting.js
'use strict'

import { lookup } from './hexagrams.js'

/**
 * 執行一次三硬幣卜卦
 * @param {import('./models.js').Hexagram[]} hexagrams - 預先載入的 64 卦資料
 * @param {string} question - 使用者問題
 * @param {() => number} [coinFn] - 回傳 2 或 3 的函式；預設以 Math.random 模擬
 * @returns {import('./models.js').CastResult}
 */
export function castHexagram(hexagrams, question, coinFn = () => (Math.random() < 0.5 ? 2 : 3)) {
  const lines = []
  for (let i = 0; i < 6; i++) {
    const sum = coinFn() + coinFn() + coinFn()
    lines.push({
      value: sum,
      yin: sum % 2 === 0,
      changing: sum === 6 || sum === 9,
    })
  }

  const lower = trigramCode(lines.slice(0, 3))
  const upper = trigramCode(lines.slice(3, 6))
  const present = lookup(hexagrams, upper, lower)

  const future = lines.some(l => l.changing)
    ? computeFuture(hexagrams, lines)
    : null

  return { question, lines, present, future }
}

/**
 * 計算三條爻的 trigram 三位碼（bit0 = 底爻，yang=1，yin=0）
 * @param {import('./models.js').Line[]} threeLines
 * @returns {number}
 */
function trigramCode(threeLines) {
  return threeLines.reduce((code, line, i) => {
    const yang = line.yin ? 0 : 1
    return code | (yang << i)
  }, 0)
}

/**
 * 計算變卦（將變爻的陰陽反轉後查卦）
 * @param {import('./models.js').Hexagram[]} hexagrams
 * @param {import('./models.js').Line[]} lines
 * @returns {import('./models.js').Hexagram}
 */
function computeFuture(hexagrams, lines) {
  const futureLines = lines.map(l => ({
    value: l.value,
    yin: l.changing ? !l.yin : l.yin,
    changing: l.changing,
  }))
  const lower = trigramCode(futureLines.slice(0, 3))
  const upper = trigramCode(futureLines.slice(3, 6))
  return lookup(hexagrams, upper, lower)
}

File: tui-go/internal/model/casting.go
package model

import "math/rand"

// CastHexagram 執行一次三硬幣卜卦
// r 為隨機源，測試時注入固定種子以確保可重現
func CastHexagram(question string, hexagrams []Hexagram, r *rand.Rand) (CastResult, error) {
	var lines [6]Line
	for i := range lines {
		sum := coin(r) + coin(r) + coin(r)
		lines[i] = Line{
			Value:    sum,
			Yin:      sum%2 == 0,
			Changing: sum == 6 || sum == 9,
		}
	}

	upper, lower := trigramCodes(lines)
	present, err := Lookup(hexagrams, upper, lower)
	if err != nil {
		return CastResult{}, err
	}

	future, err := computeFuture(hexagrams, lines)
	if err != nil {
		return CastResult{}, err
	}

	return CastResult{
		Question: question,
		Lines:    lines,
		Present:  present,
		Future:   future,
	}, nil
}

// coin 模擬一枚硬幣：正面(字面)=2，背面=3
func coin(r *rand.Rand) int {
	if r.Intn(2) == 0 {
		return 2
	}
	return 3
}

// trigramCodes 計算六爻的上下卦三位碼
// bit0 = 底爻，yang=1，yin=0
func trigramCodes(lines [6]Line) (upper, lower int) {
	yang := func(l Line) int {
		if l.Yin {
			return 0
		}
		return 1
	}
	lower = yang(lines[0]) | yang(lines[1])<<1 | yang(lines[2])<<2
	upper = yang(lines[3]) | yang(lines[4])<<1 | yang(lines[5])<<2
	return upper, lower
}

// computeFuture 計算變卦；若無變爻則回傳 nil
func computeFuture(hexagrams []Hexagram, lines [6]Line) (*Hexagram, error) {
	hasChanging := false
	for _, l := range lines {
		if l.Changing {
			hasChanging = true
			break
		}
	}
	if !hasChanging {
		return nil, nil
	}

	var futureLines [6]Line
	for i, l := range lines {
		yin := l.Yin
		if l.Changing {
			yin = !l.Yin
		}
		futureLines[i] = Line{Value: l.Value, Yin: yin, Changing: l.Changing}
	}

	upper, lower := trigramCodes(futureLines)
	h, err := Lookup(hexagrams, upper, lower)
	if err != nil {
		return nil, err
	}
	return &h, nil
}
